### ISSUE TITLE
Bumpmapping shader has incorrect behaviour and the lighting needs a new implementation

### ISSUE THREAD

**Issue by HybridDog** [minetest/minetest#9241]
##### Minetest version
```
Minetest 5.2.0-dev-37f771a9-dirty (Linux)
Using Irrlicht 1.8.4
BUILD_TYPE=Release
RUN_IN_PLACE=0
USE_GETTEXT=1
USE_SOUND=1
USE_CURL=1
USE_FREETYPE=1
USE_LUAJIT=1
STATIC_SHAREDIR="/usr/local/share/minetest"
```

##### OS / Hardware
Operating system: linux
CPU: intel

GPU model: intel
OpenGL version: 4.5

##### Summary

I cannot find an explanation about what the currently implemented bumpmapping calculates.
If I test it with normal vectors corresponding to a pyramid, it looks like this:
![2](https://user-images.githubusercontent.com/3192173/71368662-ca7acb80-25a8-11ea-9468-a9969a4b43be.png)
I am not sure if it should look like this or if it is a bug.

By the way:
For testing, I have changed code to get correct normal vectors in Tangent Space in the fragment shader.
The default lighting in minetest uses the surface normal vectors and calculates a directional ambient light with them.
I have moved the code to the fragment shader and used the normal vectors from the normalmap instead of only the surface normal: https://github.com/HybridDog/minetest/tree/m_bumpmapping_fixes
The result looks like this:
![3](https://user-images.githubusercontent.com/3192173/71369316-a61fee80-25aa-11ea-800a-92494afe5404.jpg)
There is no specular light because there are no light sources in directional ambient lighting,
so it looks quite unspectacular.

##### Steps to reproduce

Play around with bumpmapping, and the corresponding shader code

**Comment by paramat** [comment]
Good question! =D it has always looked wrong to me.

I think @lhofhansl has discovered a bug in bumpmapping, discussed in another issue somewhere.

**Comment by hecktest** [comment]
If there are no light sources to get a vector from, perhaps the vector can be guessed from the light gradient?

**Comment by HybridDog** [comment]
> perhaps the vector can be guessed from the light gradient?

The light gradient does not have information about the light source distance along the surface's normal vector. If you made an approximation, e.g. assume that this distance is zero, the lighting would probably look wrong.

It is possible to calculate a light vector to the sun, see #9272.
It is also possible to pass light positions to the shader, and then calculate light vectors for each pixel. However, a player may decide to place 16*16*16 light nodes in one mapblock, which would make lighting very slow; there exist approximations for many-light problems, http://pellacini.di.uniroma1.it/publications/lightslice11/lightslice11-paper.pdf may be related.

**Comment by hecktest** [comment]
Oh right, light in MT spreads with manhattan distance. That reduces the directions to just 8 per plane. Still, I'd like to see how it actually looks before knocking the idea completely.

The problem of many lights is best solved by deferred rendering, but you probably don't feel like touching that. Still, it's a robust solution that's the best fit for a dynamic world like MT.

The other way is to knock out lights that don't contribute, but that will always come with some popping artifacts.

**Comment by HybridDog** [comment]
> The problem of many lights is best solved by deferred rendering, but you probably don't feel like touching that.

I think that a depth prepass before rendering anything may greatly increase performance because in minetest the world consists of lots of triangles which are behind each other, e.g. caves in the underground.
I also like deferred rendering because you could add a bloom effect which is only visible in the fog.
With deferred rendering, the lights effect can be limited to the light radius, but it is probably still very slow if you have e.g. 16^3 light source nodes.

> The other way is to knock out lights that don't contribute […]

Those artifacts could be very annoying.
Another way is dynamic light map creation, but it only works for view-independent light and may be difficult to implement with animated textures. There would also be lots of light map updates because the nodes change constantly in minetest.

**Comment by paramat** [comment]
Closely related discussion:
https://github.com/minetest/minetest_game/issues/2473
https://github.com/minetest/minetest/issues/9711

**Comment by paramat** [comment]
This may be unimportant if we rework the bumpmapping light source (which i think needs doing), but how bumpmapping shading works currently seems wrong.
The screenshots below are of the new pyramid testing node, looking at all faces of a 5x5x5 cube. For cube top and base, all 4 player orientations were tested.

![screenshot_20201010_172136](https://user-images.githubusercontent.com/3686677/95660224-4fc27300-0b1e-11eb-8551-7fc921f9920a.png)

^ All cube sides appear like this. Upper and lower are correct, left/right are incorrect.

![screenshot_20201010_172353](https://user-images.githubusercontent.com/3686677/95660261-8a2c1000-0b1e-11eb-9b09-bfbf871261a8.png)

^ Cube top. Player orientated N. Upper and lower are correct, left/right are incorrect.

![screenshot_20201010_172406](https://user-images.githubusercontent.com/3686677/95660288-bf386280-0b1e-11eb-84b0-1a57bd76a192.png)

^ Cube top. Player orientated S. Upper and lower are incorrect, left/right are correct.

![screenshot_20201010_172425](https://user-images.githubusercontent.com/3686677/95660304-dc6d3100-0b1e-11eb-9821-5b66f2752886.png)

^ Cube top. Player orientated E. Upper left and lower right are correct, other diagonals incorrect.

![screenshot_20201010_172433](https://user-images.githubusercontent.com/3686677/95660334-13434700-0b1f-11eb-9d89-d6047420881a.png)

^ Cube top. Player orientated W. Upper right and lower left are correct, other diagonals incorrect.

Looking at the cube base acts like looking at the cube top.

@HybridDog are you sure the normals are set correctly in the test node? They probably are but just checking.

The bumpmapping shading seems to be intended to work as if the light source is along the player's view direction, very odd and looks wrong.

In summary:
* Bumpmapping shading seems broken.
* Even if it was fixed, it has the wrong implementation in being relative to player position and view direction.
* The light source for bumpmapping shading should be a fixed position in the world, ideally a 'directional light source' at a particular angle but at infinite distance (not sure if that is possible) so that bumpmapping appears the same way throughout a world.

**Comment by HybridDog** [comment]
> are you sure the normals are set correctly in the test node?

They should be correct. In OpenGL (0,0) is the bottom left of a texture whereas in the texture file it is the top left, so the y component of the normal vector has to be flipped: https://github.com/HybridDog/minetest/blob/90d8385033ab67570fe9944a73def118782dcb50/client/shaders/nodes_shader/opengl_fragment.glsl#L88
(I think this mirroring can also be done in the vertex shader with the transformation matrix for the Tangent Space, which is a bit more efficient.)
This mirroring may be missing in the current shaders.

**Comment by SmallJoker** [comment]
Bump mapping removed, see #10479.
